### DISCUSSION TITLE
Hello arduino + esp32

### DISCUSSION THREAD

**Discussion by leompa** [Open-Agriculture/AgIsoStack-Arduino#3] [General]
Hello, thank you very much for the new library to be used in Arduino.
A query, can be used with CAN MCP 2515 with TJA 1050. or what is the Can.
In any case, wanting to adapt the library to use this type of Canbus adaptor, which file should edit to do the communication.

**Comment by ad3154** [comment]
Hello!

For ESP32, we currently support PlatformIO with the ESP-IDF framework using [the main repo](https://github.com/Open-Agriculture/AgIsoStack-plus-plus) and the CMake build system. That version of the CAN stack supports the MCP2515 as well as the ESP32 two wire automotive interface (TWAI) built-in controller, among other things such as SocketCAN for Linux based controllers. We have [some discussion](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/discussions/225) on how to do that, and will be soon publishing a full step-by-step guide for it as well, so keep an eye out for that.

For the Arduino version, we currently only support the Teensy 4/4.1 built-in CAN controller. It's our intention to expand this to work on ESP32 with Arduino as the framework in the future, including the ESP32 TWAI controller and possibly the MCP2515 serial CAN controller, though I do not have a timeline on that at the moment.

**Reply by leompa** [reply]
Hello, perfect, I did see that there was compatibility using Platformio, but I couldn't compile. I'm going to make a clean environment, download the library with the examples and connect it to an esp32 + mcp2515. There is a tutorial or step guide to implement.

Thanks

**Reply by ad3154** [reply]
Hello again. A draft of the tutorial for PlatformIO is now available here https://delgrossoengineering.com/isobus-tutorial/Tutorials/ESP32%20PlatformIO as part of this PR on the main repository https://github.com/Open-Agriculture/AgIsoStack-plus-plus/pull/326

**Comment by DarrylMcoder** [comment]
@ad3154
All I need to do though is write a plugin to communicate with the MCP2515. Is that correct? Or is the Teensy the only one that supports the stack for some other reason? Thanks 🙏

**Reply by ad3154** [reply]
It depends...

If you're using ESP32, I really suggest using PlatformIO as [described here](https://agisostack-plus-plus.readthedocs.io/en/latest/Tutorials/ESP32%20PlatformIO.html)  with the regular CMake version of the stack instead of the Arduino version, and to use the ESP's built-in TWAI peripheral and a transceiver rather than the MCP2515 if your ESP has it. However, the non-arduino stack actually [already has an MCP2515 plugin](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/main/hardware_integration/include/isobus/hardware_integration/mcp2515_can_interface.hpp). So if you do really want to use the MCP2515, you can do it with the full version of the stack via platformIO if you want without any extra work, in theory.

If you're using something else that isn't teensy or ESP32 and want to use Arduino IDE, the main reason we don't officially list those platforms as supported is due to the sheer size of the library. Most arduino boards have between 32k and 512k of flash storage, and our library is very close to that 512k size with optimizations enabled for size. Teensy 4 has 2M, and Teensy 4.1 has a whopping 8M which seems to work OK even with a moderately large pile of data, like an ISOBUS virtual terminal object pool.

**Reply by DarrylMcoder** [reply]
That makes sense. I had been hoping to use it on the Arduino Uno with the Sparkfun CANBUS shield for starters, since that's what I have at the moment, but that is obviously impossible. Maybe I'll need some new hardware or connect the shield to my raspberry pi. 
   Anyway, keep up your good work. I haven't actually used the library yet but it looks the most promising of those I've seen.
